lib/iris/tests/unit/data_manager/test_DataManager.py

```diff
@@ -0,0 +1,602 @@
+
+    def test_scalar_1d_to_zero_ndim_fail(self):
+        dm = DataManager(np.array([123]))
+        emsg = 'Require data with shape \(1,\), got \(\).'
+        with self.assertRaisesRegexp(ValueError, emsg):
+            dm.data = 456
+
+    def test_nd_real_to_nd_real(self):
+        shape = (2, 3, 4)
+        size = np.cumprod(shape)[-1]
+        real_array = np.arange(size).reshape(shape)
+        dm = DataManager(real_array * 10)
+        self.assertTrue(dm.has_real_data())
+        self.assertFalse(dm.has_lazy_data())
+        dm.data = real_array
+        self.assertTrue(dm.has_real_data())
+        self.assertFalse(dm.has_lazy_data())
+        self.assertArrayEqual(dm.data, real_array)
+
+    def test_nd_real_to_nd_lazy(self):
+        shape = (2, 3, 4)
+        size = np.cumprod(shape)[-1]
+        real_array = np.arange(size).reshape(shape)
+        lazy_array = da.from_array(real_array, chunks=shape) * 10
+        dm = DataManager(real_array)
+        self.assertTrue(dm.has_real_data())
+        self.assertFalse(dm.has_lazy_data())
+        dm.data = lazy_array
+        self.assertFalse(dm.has_real_data())
+        self.assertTrue(dm.has_lazy_data())
+        self.assertArrayEqual(dm.data, lazy_array.compute())
+
+    def test_nd_lazy_to_nd_real(self):
+        shape = (2, 3, 4)
+        size = np.cumprod(shape)[-1]
+        real_array = np.arange(size).reshape(shape)
+        lazy_array = da.from_array(real_array, chunks=shape)
+        dm = DataManager(lazy_array * 10)
+        self.assertFalse(dm.has_real_data())
+        self.assertTrue(dm.has_lazy_data())
+        dm.data = real_array
+        self.assertTrue(dm.has_real_data())
+        self.assertFalse(dm.has_lazy_data())
+        self.assertArrayEqual(dm.data, real_array)
+
+    def test_nd_lazy_to_nd_lazy(self):
+        shape = (2, 3, 4)
+        size = np.cumprod(shape)[-1]
+        real_array = np.arange(size).reshape(shape)
+        lazy_array = da.from_array(real_array, chunks=shape)
+        dm = DataManager(lazy_array * 10)
+        self.assertFalse(dm.has_real_data())
+        self.assertTrue(dm.has_lazy_data())
+        dm.data = lazy_array
+        self.assertFalse(dm.has_real_data())
+        self.assertTrue(dm.has_lazy_data())
+        self.assertArrayEqual(dm.data, lazy_array.compute())
+
+    def test_realised_dtype_clearance(self):
+        shape = (2, 3, 4)
+        size = np.cumprod(shape)[-1]
+        mask_array = ma.arange(size).reshape(shape)
+        dtype = mask_array.dtype
+        lazy_array = as_lazy_data(mask_array)
+        dm = DataManager(lazy_array, realised_dtype=dtype)
+        self.assertEqual(dm._realised_dtype, dtype)
+        self.assertEqual(dm.dtype, dtype)
+        dm.data = mask_array
+        self.assertIs(dm.data, mask_array)
+        self.assertIsNone(dm._realised_dtype)
+        self.assertEqual(dm.dtype, dtype)
+
+    def test_coerce_to_ndarray(self):
+        shape = (2, 3)
+        size = np.cumprod(shape)[-1]
```

Comment:
If I used them in-place, then I'd be repeating the `(2, 3)` or using the `(2, 3)` and a hardwired size of `6` that is linked to the shape, opening up the possibility of a dev changing one and not the other. It's just my personal choice ... but if you feel strongly enough about it, then I'm happy to take them out.